### DISCUSSION TITLE
How to add the response header to only one endpoint

### DISCUSSION THREAD

**Discussion by Kazunari001** [Avanade/Beef#261] [Q&A] [answered]
Hi.
I'm thinking of a way to add an API response header.
However, I want to add an API response Header to only one endpoint.

So, I have come up with two ways.

The first one is to define it in `Startup.Configure`.
But this way affects all endpoints.
What I would like to discuss this time is the second way.

The second one is to create a custom API by creating a partial class in the Controller.
However, this way is outside the definition of Beef, and I'm wondering if there is a better way.

Any good ideas on this🤔

**Comment by chullybun** [comment, marked as answer]
Hi @Kazunari001, your thinking above is on the right track. _Beef_ is/was never intended to be an all-in proposition, but a means to industrialize the most common patterns. Therefore, the `Controller` class is made `partial` to enable you as the developer to extend capabilities where the _Beef_ out-of-the-box is not appropriate. 

Depending on your needs, you may still want the other layers code-generated with your custom `Controller` logic invoking. If this is the case, then add `excludeWebApi: true` to the `operation` code-gen configuration. _FYI:_ there is documentation for all of the possible code-gen configurations; see [here](https://github.com/Avanade/Beef/blob/master/docs/Entity-CodeGeneration-Config.md).

Finally, this is a bit more hacky but is another means to achieve your goal. Version [`5.8.0`](https://github.com/Avanade/Beef/blob/master/CHANGELOG.md#v580) introduced support for a return value of `IActionResult`. The generated `Controller`-based _CoreEx_ [`WebApi`](https://github.com/Avanade/CoreEx/tree/main/src/CoreEx.AspNetCore#limitations) logic will then pass through as-is achieving whatever result processing you desire. This was originally intended for the likes of [`FileContentResult`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.filecontentresult); but is flexible enough for anything. Also, _CoreEx's_ [`ExtendedContentResult`](https://github.com/Avanade/CoreEx/blob/main/src/CoreEx.AspNetCore/WebApis/ExtendedContentResult.cs) or [`ValueContentResult`](https://github.com/Avanade/CoreEx/blob/main/src/CoreEx.AspNetCore/WebApis/ValueContentResult.cs) may prove useful to enable whatever `HttpResponse` handling is required.

Personally, I prefer the first version, as that avoids having _ASP.NET Web Api_ types used inside of business logic, and the intent within the `Controller` is more explicit. I.e. you create your `IActionResult` object in the `Controller` and return it. I have just added an example; have a look [here](https://github.com/Avanade/Beef/blob/master/samples/Demo/Beef.Demo.Api/Controllers/PersonController.cs) with its unit test [here](https://github.com/Avanade/Beef/blob/master/samples/Demo/Beef.Demo.Test/PersonTest.cs) (scroll to bottom).

Hope this helps,
Cheers 👊

**Reply by Kazunari001** [reply]
Hi @chullybun ,
Thank you so much for your detailed response, and for taking the time to create a sample as well.
I fully agree with your explanation, and after reviewing the sample you provided, I'm excited to continue exploring Beef further.👊
